src/modules/Elsa.Workflows.Core/Services/ActivityRegistryUpdateService.cs

```diff
@@ -0,0 +1,54 @@
+using Elsa.Workflows.Contracts;
+using Elsa.Workflows.Models;
+
+namespace Elsa.Workflows.Services;
+
+/// <summary>
+/// Service responsible for updating the activity registry based on activity providers.
+/// </summary>
+public class ActivityRegistryUpdateService(IEnumerable<IActivityProvider> providers, IActivityRegistry registry) : IActivityRegistryUpdateService
+{
+    /// <inheritdoc />
+    public async Task AddToRegistry(Type providerType, string workflowDefinitionVersionId, CancellationToken cancellationToken = default)
+    {
+        var provider = providers.First(x => x.GetType() == providerType);
+        var descriptors = await provider.GetDescriptorsAsync(cancellationToken);
+        var descriptorToAdd = descriptors
+            .SingleOrDefault(d =>
+                d.CustomProperties.TryGetValue("WorkflowDefinitionVersionId", out var val) &&
+                val.ToString() == workflowDefinitionVersionId);
+        
+        if (descriptorToAdd is not null)
+            registry.Add(providerType, descriptorToAdd);
```

Comment:
Keeping my previous comment in mind, this line would become `registry.Add(typeof(WorkflowDefinitionActivityProvider), descriptorToAdd)`.